core/src/main/java/com/linecorp/armeria/common/util/Unwrappable.java

```diff
@@ -87,4 +87,41 @@ default <T> T as(Class<T> type) {
     default Unwrappable unwrap() {
         return this;
     }
+
+    /**
+     * Unwraps this object and returns the object being decorated. If {@code all} is true,
+     * then the innermost wrapped object is returned. If {@code all} is false, then the behavior is identical
+     * to {@link Unwrappable#unwrap()}. If this {@link Unwrappable} is the innermost
+     * object, this method returns itself. For example:
+     * <pre>{@code
+     * class Foo implements Unwrappable {}
+     *
+     * class Bar<T extends Unwrappable> extends AbstractUnwrappable<T> {
+     *     Bar(T delegate) {
+     *         super(delegate);
+     *     }
+     * }
+     *
+     * class Qux<T extends Unwrappable> extends AbstractUnwrappable<T> {
+     *     Qux(T delegate) {
+     *         super(delegate);
+     *     }
+     * }
+     *
+     * final Foo foo = new Foo();
+     * foo.unwrap(true) == foo;
+     * foo.unwrap(false) == foo;
+     *
+     * final Bar<Foo> bar = new Bar<>(foo);
+     * bar.unwrap(true) == foo;
+     * bar.unwrap(false) == foo;
+     *
+     * final Qux<Bar<Foo>> qux = new Qux<>(bar);
+     * qux.unwrap(true) == foo;
+     * qux.unwrap(false) == bar;
+     * }</pre>
+     */
+    default Unwrappable unwrap(boolean all) {
```

Comment:
- use `unwrap()` instead of `this` because some implementations directly return the inner object.
https://github.com/line/armeria/blob/5ed3cacc8e5a5ec14e5b850e4c23b260b0468334/core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java#L194-L197
- Should we override/implement this method in some classes where `AbstractUnwrappable` isn't used?
https://github.com/line/armeria/blob/435aeedcb007382250e4ee71311a5ff2c56615ae/core/src/main/java/com/linecorp/armeria/client/DefaultBlockingWebClient.java#L82-L85
- Should we change the return type of this method where possible?
https://github.com/line/armeria/blob/eb33d2b77fb9f600b3eee14d4c601fdef32add6f/core/src/main/java/com/linecorp/armeria/client/Client.java#L85

Reply:
> How about using unwrap() instead of this because some implementations directly return the inner object?

Good point 👍  Done.

> Should we override/implement this method in some classes where AbstractUnwrappable isn't used?

Good point. I think `unwrapAll` and `unwrap` should work consistently (as in if `unwrap()` is continuously called, it should reach the same value as `unwrapAll()`). Unfortunately, it seems difficult to find a way to automatically verify/enforce this constraint.

>Should we change the return type of this method where possible?

I actually didn't do this intentionally since it's not possible to know the return type of `unwrapAll` from `AbstractUnwrappable`.

Another way I thought we could do this is by unsafe casting the return type. I decided against this approach since I was concerned the loss of type safety, but let me know what you think.
```
public <U extends Unwrappable> U unwrapAll() {
    return (U) this;
}
```

ref: https://github.com/line/armeria/pull/4338#discussion_r914680171

Reply:
> I actually didn't do this intentionally since it's not possible to know the return type of unwrapAll from AbstractUnwrappable.

I see. Let's keep it as is.